packages/jest-environment-jsdom/src/index.ts

```diff
@@ -160,13 +160,6 @@ export default class JSDOMEnvironment implements JestEnvironment<number> {
         this.global.removeEventListener('error', this.errorEventListener);
       }
       this.global.close();
-
-      // Dispose "document" to prevent "load" event from triggering.
-
-      // Note that this.global.close() will trigger the CustomElement::disconnectedCallback
-      // Do not reset the document before CustomElement disconnectedCallback function has finished running,
-      // document should be accessible within disconnectedCallback.
-      Object.defineProperty(this.global, 'document', {value: null});
```

Comment:
Would this fix https://github.com/facebook/jest/issues/12670? Worth adding a test?

Reply:
quite possibly...

Reply:
https://github.com/eps1lon/react-testing-library-error-repro/tree/upgrade-jsdom now errors with 

```
Error: The `document` global was defined when React was initialized, but is not defined anymore. This can happen in a test environment if a component schedules an update from an asynchronous callback, but the test has already finished running. To solve this, you can either unmount the component at the end of your test (and ensure that any asynchronous operations get canceled in `componentWillUnmount`), or you can change the test itself to be asynchronous.
```

when using jest-environment-jsdom. This is certainly a clearer error but still unintended.

Going to do some more digging.

Reply:
Yeah, I sent https://github.com/facebook/react/pull/22695 which fixed the check for the correct error

Reply:
Hi @SimenB I started looking at this upgrade PR, because many of the [JSDOM@21.1.0](https://github.com/jsdom/jsdom/blob/master/Changelog.md#2110) patches are mine and I would like to start using them in our Jest tests ASAP 🙂 

I did some research about why the `document` is being set to `null` during teardown, and why that doesn't really need to be done, and summarized that in #13972.

Reply:
> `react-testing-library-error-repro` now errors with

@eps1lon I think this is not an issue with Jest, but a bug in React 17's `act` implementation, when the `act` callback returns a promise. Then, after the promise resolves, React will synchronously flush all updates scheduled before resolving. But that code is buggy: it leaves the `act` environment _before_ doing that flush, so these updates no longer work with the `act` queue, but are using the regular scheduler. `setImmediate` etc. And then they run long after the JSDOM env has been teared down.

React 18 has that fixed: it will flush the updates while still inside the `act` environment.

In your repro example, upgrading to React 18 in `package.json` immediately fixes the test for me.